master/internal/provisioner/aws_spot.go

```diff
@@ -0,0 +1,667 @@
+	activeSpotRequests      map[string]*spotRequest
+	onlyLogErrorOnceTracker map[string]bool
+	inBackoffState          bool
+	backoffStart            time.Time
+	approximateClockSkew    *time.Duration
+	launchTimeOffset        time.Duration
+}
+
+func (c *awsCluster) listSpot(ctx *actor.Context) ([]*Instance, error) {
+	// This list operation will become slower as the number of spot requests grows. Spot Instance
+	// requests are deleted from the AWS API four hours after they are canceled and their instances
+	// are terminated. In normal usage this shouldn't be a performance issue, but it's important we
+	// avoid endlessly resubmitting requests that will instantly fail.
+	spotRequests, err := c.listSpotInstanceRequests(ctx, false)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot describe EC2 spot requests")
+	}
+
+	activeRequests, inactiveRequests := c.splitRequestsIntoActiveAndInactive(ctx, spotRequests)
+
+	ctx.Log().
+		WithField("log-type", "listSpot.querySpotRequest").
+		Debugf("Retrieved spot requests: %d active requests and %d inactive requests",
+			len(activeRequests), len(inactiveRequests))
+
+	c.updateBackoffState(ctx, spotRequests)
+	c.cleanup(ctx, spotRequests)
+	c.updateActiveRequestSnapshot(ctx, activeRequests, inactiveRequests)
+	instances, err := c.buildInstanceListFromActiveRequestSnapshot(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return instances, nil
+}
+
+func (c *awsCluster) terminateSpot(ctx *actor.Context, instanceIDs []*string) {
+	if len(instanceIDs) == 0 {
+		return
+	}
+
+	instancesToTerminate := make([]*string, 0, 0)
+	pendingSpotRequestsToTerminate := make([]*string, 0, 0)
+
+	for _, instanceId := range instanceIDs {
+		if strings.HasPrefix(*instanceId, spotRequestIdPrefix) {
+			spotRequestId := instanceId
+			pendingSpotRequestsToTerminate = append(pendingSpotRequestsToTerminate, spotRequestId)
+		} else {
+			instancesToTerminate = append(instancesToTerminate, instanceId)
+		}
+	}
+
+	ctx.Log().Debugf(
+		"terminating %d EC2 instances and %d spot requests: %s,  %s",
+		len(instancesToTerminate),
+		len(pendingSpotRequestsToTerminate),
+		instancesToTerminate,
+		pendingSpotRequestsToTerminate,
+	)
+
+	terminateInstancesResponse, err := c.terminateInstances(instancesToTerminate, false)
+	if err != nil {
+		ctx.Log().WithError(err).Error("cannot terminate EC2 instances")
+	} else {
+		terminated := c.newInstancesFromTerminateInstancesOutput(terminateInstancesResponse)
+		ctx.Log().Debugf(
+			"terminated %d EC2 instances: %s",
+			len(terminated),
+			fmtInstances(terminated),
+		)
+	}
+
+	_, err = c.terminateSpotInstanceRequests(ctx, pendingSpotRequestsToTerminate, false)
+	if err != nil {
+		ctx.Log().WithError(err).Error("cannot terminate spot requests")
+	} else {
+		ctx.Log().Debugf(
+			"terminated %d spot requests: %s",
+			len(pendingSpotRequestsToTerminate),
+			pendingSpotRequestsToTerminate,
+		)
+	}
+
+	// TODO: Race condition - an instance could have been created between listing
```

Comment:
I think it's fine to just remove this comment. This type of race condition happens for other providers since we can only react to a snapshot of instances since AWS/GCP has the only source of truth.